Support installation from package.json file

nihey opened the pull request:
Even though it was suggested modifying `Gpm`'s `pack` on https://github.com/IonicaBizau/node-gpm/issues/2. I thought It could be nice to extend the API to accept packages via string or list.

So I changed `Gpm.prototype.install` function to handle a list of packages, and moved the old `Gpm.prototype.install` to `Gpm.prototype.installCurrent`, the colateral effect of that is moving the url parsing from the `constructor` to the `Gpm.prototype.installCurrent` function. I hope that's alright with you.

nihey committed:
Support installation from package.json file

IonicaBizau commented:
Sounds good! :+1: 

Please [take a look at the code style](https://github.com/IonicaBizau/code-style) and [add yourself as contributor](https://github.com/IonicaBizau/node-gpm/blob/master/CONTRIBUTING.md#fixing-issues).

Thanks! :sparkle:

nihey commented:
Alright :)

It seems I forgot the contributor part.

On the code style, I tried to follow it all, have I missed something?

Edit: Just seen it, the double quotes, I'll fix it.

nihey committed:
Add Nihey Luz Takizawa to the contributors.

IonicaBizau commented on bin/gpm:
```diff
@@ -53,6 +54,17 @@ if (destOpt.value) {
 
+if (inputOpt.value === ".") {
+    // Load and parse the package.json file
+    var packageJsonPath = Path.join(process.cwd(), "package.json");
```
Please use the following style:

``` js
var foo = ...
  , bar = 42
  ;
```

nihey replied:
Oh, now I see, sorry about that, it shall be fixed.

IonicaBizau commented on bin/gpm:
```diff
@@ -9,6 +9,7 @@ var Gpm = require("../lib")
   , Deffy = require("deffy")
   , Ul = require("ul")
   , Path = require("path")
+  , Fs = require("fs")
```
Use `ReadJson = require("r-json")` instead.

IonicaBizau commented on bin/gpm:
```diff
@@ -53,6 +54,17 @@ if (destOpt.value) {
+if (inputOpt.value === ".") {
+    // Load and parse the package.json file
+    var packageJsonPath = Path.join(process.cwd(), "package.json")
+      , manifest = JSON.parse(Fs.readFileSync(packageJsonPath));
```
And here use `ReadJson(packageJsonPath)`. Also, put the `;` on a new line.

IonicaBizau commented on bin/gpm:
```diff
@@ -53,6 +54,17 @@ if (destOpt.value) {
+    // Merge all dependencies on a single Array
+    var dependencies = Object.keys(manifest.dependencies)
+      , devDependencies = Object.keys(manifest.devDependencies);
+    inputOpt.value = dependencies.concat(devDependencies);
```
Is it possible to add a new method for this? Something like `Gpm.prototype.getDependencies = ...`, and considering the `is_dev` field (`self.is_dev`).

nihey replied:
Sure, I just have some questions about the `is_dev` field, do you want me to:
- Add `devDependencies` only when `is_dev` is true.
- or
- Set `is_dev` for each package individually depending on whether it is on `devDependencies` or not.

On the first scenario, I might have to read the `package.json` inside the `Gpm`, as `Gpm` currently don't know which dependencies are devDependencies - and it is not instanced yet.

On the second scenario, `Gpm` will have to accept an `Array` of `Object`s to properly describe which dependencies are dev (this could also be used for the first scenario).

*I'll fix the `;` on the new line here too.

IonicaBizau replied:
`is_dev` can be set when the `Gpm` instance is initialized:

``` js
var pack = new Gpm("foo", { is_dev: true });
```

That means `Gpm.install` will install dev dependencies as well, recursively (will initialize `Gpm` using `is_dev:true`).

So, would be nice to have something like this:

``` js
console.log(pack.getDependencies());
```

This will output: `["foo", "bar", "baz"]` or `["foo", "bar", "baz", "mocha"]`, if `is_dev:true`.

Hope you got it! Ask me if you have something that is not clear. Making this method async is not a problem.

On the other side, I'm not really happy with accepting a list of strings in `Gpm` constructor. It's better to handle that in the bin script and calling `dep.install()` for each of them.

Anyway, I'm sure it will be great! :smile:

nihey replied:
Understood :smile:, I'll do it.

I should also point that I think [this line](https://github.com/IonicaBizau/node-gpm/blob/master/lib/index.js#L217) should actually be testing `self.is_dev`.

IonicaBizau replied:
> I should also point that I think this line should actually be testing self.is_dev.

Whoups, true! Indeed! :+1:

IonicaBizau commented on lib/index.js:
```diff
@@ -30,11 +30,11 @@ var el = new ExecLimiter(Deffy(parseInt(process.env.GPM_PROCESS_COUT_LIMIT) || 1
  * @name Gpm
  * @function
- * @param {String} name The git url or the npm package name.
+ * @param {(String|String[])} packages The git url, the npm package name or a list of them.
```
I was using `{String|Array}` -- it's obvious that it's an array of strings. :smile:

IonicaBizau commented on lib/index.js:
```diff
@@ -43,19 +43,15 @@ function Gpm(name, options) {
       , is_dev: false
     });
 
+    if (typeof(packages) === "string") {
```
Can you please change this to `typeof packages === "string"`? I have never used `typeof` as function.
Maybe I have to add that to code stlye.

IonicaBizau commented:
@nihey Few more things to fix. :smile: It's not mandatory to put all the changes in one commit, just commit any times it's needed.

nihey committed:
Fix some coding style errors

nihey committed:
Add getDependencies method to Gpm

- Packages that require installation are now listed by getDependencies method
- Drop support to list of packages on Gpm.
- Add support for a package.json object on Gpm.

nihey commented:
I tried to make some structural changes from the previous version. I made possible for `Gpm` to load a `package.json`-like object, so that `Gpm.prototype.getDependencies` could be made.

Dependency installing now considers the `is_dev` option. :smile: 

edit: I can't comment on the diff, so I'll drop the observations down here:

I also made some tests regarding the `self.dev_install` fix:
- Not all packages contain a `devDependencies` field ([example](https://github.com/IonicaBizau/node-flat-colors/blob/master/package.json)). If supporting such packages is desired, [this line](https://github.com/IonicaBizau/node-gpm/blob/master/lib/index.js#L218) should be reconsidered (or the object may assume a default value on [this line](https://github.com/IonicaBizau/node-gpm/blob/master/lib/index.js#L204)).
- The [recursive magic](https://github.com/IonicaBizau/node-gpm/blob/master/lib/index.js#L202) may be dangerous, I noticed some projects have circular dependencies (Specially circular devDependencies).

Regarding the complexity of this issues I believe fixing the `self.dev_install` functionality may be something to be done on another patch (if it will be done)

edit2: Example of circular dev dependency:

[mocha](https://github.com/mochajs/mocha/blob/master/package.json#L57) -> [coffee-script](https://github.com/jashkenas/coffeescript/blob/master/package.json#L39) -> [underscore](https://github.com/jashkenas/underscore/blob/master/package.json#L23) -> [karma](https://github.com/karma-runner/karma/blob/master/package.json#L260) -> mocha

IonicaBizau commented:
Circular dependencies is, indeed, an issue. I will take a look soon on what you have done!
Thanks a lot! :sparkles:

IonicaBizau commented:
An issue I see is that doing `gpm -i .` doesn't create the `node_modules` directory.

nihey commented:
You're right, I was currently handling that by using `$ gpm -i . -d node_modules/`. Maybe I should default the destination path to `node_modules` when the user has not set it by himself?

IonicaBizau commented:
> Maybe I should default the destination path to node_modules when the user has not set it by himself?

Exactly!

nihey committed:
Default the destination path to node_modules when installing from package.json

IonicaBizau commented on bin/gpm:
```diff
@@ -53,6 +54,12 @@ if (destOpt.value) {
+if (inputOpt.value === ".") {
+    // Load a package.json file in the current directory
+    inputOpt.value = ReadJson(Path.join(process.cwd(), "package.json"));
+    destOpt.value = destOpt.value || Path.join(process.cwd(), "node_modules");
```
Almost there!

``` js
destOpt.value = Deffy(destOpt.value, "node_modules");
```

And move this `if` before the previous one (line 53). That will help us reusing the same code.

nihey replied:
You're right, I just missed that.

nihey committed:
Make use of Path.resolve when defining the default destination path

IonicaBizau commented:
Finally landed on `master`. :rocket:
